labs/passport-adapter/src/__tests__/acceptance/authentication-with-passport-adapter.acceptance.ts

```diff
@@ -0,0 +1,201 @@
+import { PassportAdapter } from '../../';
+const SequenceActions = RestBindings.SequenceActions;
+const AUTH_STRATEGY_NAME = 'basic';
+
+describe('Basic Authentication - passport', () => {
+  let app: Application;
+  let server: RestServer;
+  let users: UserRepository;
+  beforeEach(givenAServer);
+  beforeEach(givenUserRepository);
+  beforeEach(givenControllerInApp);
+  beforeEach(givenAuthenticatedSequence);
+
+  it('authenticates successfully for correct credentials', async () => {
+    const client = whenIMakeRequestTo(server);
+    const credential =
+      users.list.joe.profile.id + ':' + users.list.joe.password;
+    const hash = Buffer.from(credential).toString('base64');
+    await client
+      .get('/whoAmI')
+      .set('Authorization', 'Basic ' + hash)
+      .expect(users.list.joe.profile.id);
+  });
+
+  it('returns error for invalid credentials', async () => {
+    const client = whenIMakeRequestTo(server);
+    const credential = users.list.Simpson.profile.id + ':' + 'invalid';
+    const hash = Buffer.from(credential).toString('base64');
+    await client
+      .get('/whoAmI')
+      .set('Authorization', 'Basic ' + hash)
+      .expect(401);
+  });
+
+  it('allows anonymous requests to methods with no decorator', async () => {
+    class InfoController {
+      @get('/status')
+      status() {
+        return { running: true };
+      }
+    }
+
+    app.controller(InfoController);
+    await whenIMakeRequestTo(server)
+      .get('/status')
+      .expect(200, { running: true });
+  });
+
+  function givenUserRepository() {
+    users = new UserRepository({
+      joe: { profile: { id: 'joe' }, password: '12345' },
+      Simpson: { profile: { id: 'sim123' }, password: 'alpha' },
+      Flinstone: { profile: { id: 'Flint' }, password: 'beta' },
+      George: { profile: { id: 'Curious' }, password: 'gamma' },
+    });
+  }
+
+  class PassportBasicAuthProvider implements Provider<AuthenticationStrategy> {
+    value(): AuthenticationStrategy {
+      const basicStrategy = this.configuredBasicStrategy(verify);
+      // Ideally the passport adapter should accept a configured
+      // passport instance like:
+
+      // const passport = new Passport();
+      // passport.use(basicStrategy);
+      // return new PassportAdapter(passport);
+
+      // The workaround still takes a strategy because the types exported from
+      // `passport` module gives error when using it as a type:
+      // afunction(passport: Passport){}
+      // the code above has type error. I've tried all the possible types exported
+      // from `passport` module and none of them work.
+      return new PassportAdapter(basicStrategy, 'basic');
+    }
+
+    configuredBasicStrategy(verifyFn: BasicVerifyFunction): BasicStrategy {
+      return new BasicStrategy(verifyFn);
+    }
+  }
+
+  function verify(username: string, password: string, cb: Function) {
+    process.nextTick(() => {
```

Comment:
Why do we need to wrap it into `process.nextTick()`? The `users.find` is already async.